data/src/main/java/me/androidbox/data/worker_manager/TaskCreateSynchronizerWorker.kt

```diff
@@ -0,0 +1,78 @@
+package me.androidbox.data.worker_manager
+
+import android.content.Context
+import androidx.hilt.work.HiltWorker
+import androidx.work.CoroutineWorker
+import androidx.work.WorkerParameters
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedInject
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import me.androidbox.data.local.dao.TaskDao
+import me.androidbox.data.local.entity.TaskEntity
+import me.androidbox.data.mapper.toTaskDto
+import me.androidbox.data.remote.network.task.TaskService
+import me.androidbox.data.remote.util.CheckResult.checkResult
+import me.androidbox.data.worker_manager.constant.Constant.RETRY_COUNT
+import me.androidbox.domain.constant.SyncAgendaType
+
+@HiltWorker
+class TaskCreateSynchronizerWorker @AssistedInject constructor(
+    @Assisted private val context: Context,
+    @Assisted private val workerParameters: WorkerParameters,
+    private val taskService: TaskService,
+    private val taskDao: TaskDao
+) : CoroutineWorker(context, workerParameters) {
+
+    override suspend fun doWork(): Result {
+        val createdTasks: Deferred<List<TaskEntity>>
+
+        /** Get all the created tasks */
+        supervisorScope {
+            createdTasks = async {
+                taskDao.getAllTasksBySyncType(SyncAgendaType.CREATE).map {
+                    it.task
+                }
+            }
+        }
+
+        val result = if (createdTasks.await().isNotEmpty()) {
+            if(runAttemptCount > RETRY_COUNT) {
+                Result.failure()
+            }
+            else {
+                Result.retry()
+            }
+
+            val responseState = checkResult{
+                val createdTasksEntity = createdTasks.await()
```

Comment:
Theres a big issue here

Lets say you have 10 tasks that need sync

If 8 go well, and 2 fail, you will return a failure, and still have 10 in the db as not synced. When in fact, 8 were successfuly synced

The issue will happen when 
1- Some tasks need to be created
2- The worker fails (because another failed)
3- When retrying, the creation will be done again. And it will fail since its a creation on an already created task. Meaning the worker will fail again... And again... And again. So this worker will never work

You should change the sync stuff on each item individually

Reply:
@MKiperszmid If I understand what you are saying. It will loop through and call `/create`  endpoint for each task that needs to be sync'ed. If it fails on one of them. The worker will fail and retry on all of them again? And as some of them has already been created already it will fail on the remote for those ones again. 

I am not sure I fully understand your solution. I think you might mean this:

```
        if (createdTasks.isNotEmpty()) {
            val responseState = checkResult{

                supervisorScope {
                    launch {
                        createdTasks.map { taskEntity ->
                            /* Create the task on remote */
                            taskService.createTask(taskEntity.toTaskDto())
                            /* If success will run this delete - else if loop next next one */
                            taskDao.deleteSyncTaskById(taskEntity.id)
                        }
                    }
                }
            }

            responseState.fold(
                onSuccess = {
                    /* Would this be nessecary if the items have already been deleted in the above code  */
                    taskDao.deleteSyncTasksBySyncType(SyncAgendaType.CREATE)
                    Result.success()
                },
                onFailure = {
                    Result.retry()
                }
            )
        }
```
If this is not correct can you please provide a code snippet for greater clarity?

Reply:
Yes this looks better to me, good catch btw @MKiperszmid. Basically you need to retrieve all tasks at the beginning of the worker that need to be synced, but if some have been synced successfully before, they won't be contained in the next run of this worker if you've deleted them from the table before